runtime-version/main.go

```diff
@@ -0,0 +1,142 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/spf13/cobra"
+	"google.golang.org/grpc"
+
+	"github.com/oasisprotocol/oasis-core/go/common"
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature"
+	consensusAPI "github.com/oasisprotocol/oasis-core/go/consensus/api"
+	genesisFile "github.com/oasisprotocol/oasis-core/go/genesis/file"
+	cmdCommon "github.com/oasisprotocol/oasis-core/go/oasis-node/cmd/common"
+	cmdCommonFlags "github.com/oasisprotocol/oasis-core/go/oasis-node/cmd/common/flags"
+	cmdGrpc "github.com/oasisprotocol/oasis-core/go/oasis-node/cmd/common/grpc"
+	registryAPI "github.com/oasisprotocol/oasis-core/go/registry/api"
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+
+	metadataRegistry "github.com/oasisprotocol/metadata-registry-tools"
+)
+
+var queryCmd = &cobra.Command{
+	Use:   "<runtime-id>",
+	Short: "query runtime versions",
+	Run:   doQuery,
+}
+
+var runtimeID common.Namespace
+
+func doConnect(cmd *cobra.Command) *grpc.ClientConn {
+	if err := cmdCommon.Init(); err != nil {
+		cmdCommon.EarlyLogAndExit(err)
+	}
+
+	conn, err := cmdGrpc.NewClient(cmd)
+	if err != nil {
+		cmdCommon.EarlyLogAndExit(err)
+	}
+
+	return conn
+}
+
+func doQuery(cmd *cobra.Command, args []string) {
+	ctx := context.Background()
+
+	genesis, err := genesisFile.DefaultFileProvider()
+	if err != nil {
+		cmdCommon.EarlyLogAndExit(err)
+	}
+	doc, err := genesis.GetGenesisDocument()
+	if err != nil {
+		cmdCommon.EarlyLogAndExit(err)
+	}
+	doc.SetChainContext()
+
+	if len(args) != 1 {
+		cmdCommon.EarlyLogAndExit(fmt.Errorf("need exactly one argument (runtimeID)"))
+	}
+	if err = runtimeID.UnmarshalHex(args[0]); err != nil {
+		cmdCommon.EarlyLogAndExit(fmt.Errorf("malformed runtime ID: %s", args[0]))
+	}
+	conn := doConnect(cmd)
+
+	consensus := consensusAPI.NewConsensusClient(conn)
+	reg := registryAPI.NewRegistryClient(conn)
+
+	// Take height from latest block
+	blk, err := consensus.GetBlock(ctx, consensusAPI.HeightLatest)
+	if err != nil {
+		cmdCommon.EarlyLogAndExit(err)
+	}
+	height := int64(blk.Height)
+
+	// Get nodes
+	nodes, err := reg.GetNodes(ctx, height)
+	if err != nil {
+		cmdCommon.EarlyLogAndExit(err)
+	}
+
+	entityVersions := make(map[signature.PublicKey]string)
+	versionCounts := make(map[string]int)
+	totalNodes := 0
+
+	for _, node := range nodes {
+		for _, runtime := range node.Runtimes {
+			if runtime.ID == runtimeID {
+				versionString := runtime.Version.String()
+				versionCounts[versionString] = versionCounts[versionString] + 1
+				totalNodes += 1
+
+				// todo: broken if an entity has multiple nodes running different versions
```

Comment:
Is there an easy fix for this? Otherwise, maybe just print a warning that this occurred.

Reply:
Yeah, just keep a list of versions per entity.

Reply:
Ok, I've added a map from version strings to bools.